src/Compilers/CSharp/Portable/Errors/LazyMissingNonNullTypesContextDiagnosticInfo.cs

```diff
@@ -10,49 +11,69 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// </summary>
     internal sealed class LazyMissingNonNullTypesContextDiagnosticInfo : LazyDiagnosticInfo
     {
-        private readonly CSharpCompilation _compilation;
-        private readonly bool _isNullableEnabled;
         private readonly TypeSymbolWithAnnotations _type;
+        private readonly DiagnosticInfo _info;
 
-        internal LazyMissingNonNullTypesContextDiagnosticInfo(CSharpCompilation compilation, bool isNullableEnabled, TypeSymbolWithAnnotations type)
+        private LazyMissingNonNullTypesContextDiagnosticInfo(TypeSymbolWithAnnotations type, DiagnosticInfo info)
         {
             Debug.Assert(!type.IsNull);
-            _compilation = compilation;
-            _isNullableEnabled = isNullableEnabled;
             _type = type;
+            _info = info;
         }
 
-        protected override DiagnosticInfo ResolveInfo()
+        public static void AddAll(bool isNullableEnabled, TypeSymbolWithAnnotations type, Location location, DiagnosticBag diagnostics)
         {
-            return ReportNullableReferenceTypesIfNeeded(_compilation, _isNullableEnabled, _type);
+            var rawInfos = ArrayBuilder<DiagnosticInfo>.GetInstance();
+            GetRawDiagnosticInfos(isNullableEnabled, (CSharpParseOptions)location.SourceTree.Options, rawInfos);
+            foreach (var rawInfo in rawInfos)
+            {
+                diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(type, rawInfo), location);
+            }
+            rawInfos.Free();
+        }
+
+        private static void GetRawDiagnosticInfos(bool isNullableEnabled, CSharpParseOptions options, ArrayBuilder<DiagnosticInfo> infos)
```

Comment:
>infos [](start = 131, length = 5)

Do we anticipate needing to add more diagnostics here? Seems like it might be better to just return a tuple of the two DiagnosticInfos that can be returned here, rather than going to the effort of allocating an ArrayBuilder and ensuring that it's freed afterwards.

Reply:
This leaves the door open to adding any number of diagnostics, which I think is worth it. The allocations are pooled, so I'm not worried about that.